src/nfa/operations.cc

```diff
@@ -249,7 +249,7 @@ namespace {
             }
 
             // add moves of S to the sync ex iterator
-            // TODO: shouldn't we also reset first?
+            synchronized_iterator.reset();
```

Comment:
What happened here? Is this needed?

Reply:
Even after looking at it previously, I still cannot explain how it managed to work without the reset. One should intuitively clear the iterator for every macrostate. `clear()` is a fast O(1) operation, so I added it as this is what I would do if I implemented the function again. I can take a look once more and probably try to debug it to make sure.

Reply:
Aha, I have done some more digging and I got it now. It works thanks to a lucky coincidence. Since we have been up until now always using the `SynchronizedIterator` (called advancer further for clarity) only to fully iterate over the whole vectors iterated over, the advancer remains clear after the while loops `while(iterator.advance())`. That holds only because at the end of every `advance()`, the iterators kept in the advancer are popped from the advancer when they iterated all the way to their respective end iterators. If we were to stop the iteration earlier for any macrostate, the iterators would remain in the advancer and the next macrostate would only add new iterators to these iterators from the previous macrostate. The advancer would iterate over all of them (the remaining ones and the new one which one actually want to iterate over) together, causing the `advance()` to not work correctly.

As a rule of thumb, we should therefore always clear our advancers to make sure that this happenstance cannot occur. Henceforth, I consider this change to add `clear()` everywhere a valid one.